File: lists/arrayList/c/strArrayList.c
//
// strArrayList.c
// By Scott Hall
//
// Models the data structure for a string type ArrayList

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "strArrayList.h"

// An initializer (sort of a constructor) for string array list
void StrArrayList(int size, int maxStrLength, strArrayList *arrayList) {

    int i;

    arrayList->maxSize      = size;
    arrayList->maxStrLength = maxStrLength;
    arrayList->length       = 0;
    arrayList->aList        = (char **) calloc(size, sizeof(char *));

    for (i=0; i<size; i++) {
      arrayList->aList[i] = (char *) calloc(maxStrLength, sizeof(char));
    }

}

//-----------------------------------------------------------------------------

// Return the maximum allowed size of the ArrayList
int getMaxSize(strArrayList arrayList) {
    return arrayList.maxSize;
}

//-----------------------------------------------------------------------------

// Return the current length of the array list
int getLength(strArrayList arrayList) {
    return arrayList.length;
}

// --------------------------------------------------------------------------

// Returns whether the list is empty or not
int isEmpty(strArrayList arrayList) {
    if (arrayList.length == 0) {
        return 1;
    }
    return 0;
}

// --------------------------------------------------------------------------

// Returns whether the list is full
int isFull(strArrayList arrayList) {
    if (arrayList.length == arrayList.maxSize) {
        return 1;
    }
    return 0;
}

// --------------------------------------------------------------------------

// Clear the ArrayList
void clear(strArrayList arrayList) {
    arrayList.length = 0;
}

// --------------------------------------------------------------------------

// Check if a specified item is in the ArrayList.  Returns the index of the
// match or a -1 if the item is not in the list
int searchArrayList(char *str, strArrayList arrayList) {
    int i;

    for (i=0; i<arrayList.length; i++) {
        if (strcmp(str,arrayList.aList[i]) == 0) {
            return i;
        }
    }
    return -1;
}

// --------------------------------------------------------------------------

// Insert an item at the end of the list if it does not already exist
int insertAtEnd(char *str, strArrayList *arrayList) {

    if (isFull(*arrayList) == 1) {
        printf("insertAtEnd Error: List is already full.  Cannot add new item\n");
        return 1;
    }

    if (searchArrayList(str, *arrayList) != -1) {
        printf("insertAtEnd Error: List is already has '%s'.  Cannot add new item\n", str);
        return 1;
    }

    strcpy(arrayList->aList[arrayList->length], str);
    arrayList->length++;

    return 0;
}

// --------------------------------------------------------------------------

// Insert an item at a specified index.  Index must exist, item not already
// exist and the list is not full for item to be added.  All items are
// shifted
int insertAtIndex(int index, char *str, strArrayList *arrayList) {

    int i;

    if (isFull(*arrayList) == 1) {
        printf("insertAtEnd Error: List is already full.  Cannot add new item\n");
        return 1;
    }

    if (searchArrayList(str, *arrayList) != -1) {
        printf("insertAtEnd Error: List is already has '%s'.  Cannot add new item\n", str);
        return 1;
    }

    if (index > arrayList->length || index < 0) {
        printf("insertAtEnd Error: Index is out of range\n");
        return 1;
    }

    arrayList->length++;

    for (i=arrayList->length; i>=index; i--) {
        strcpy(arrayList->aList[i+1], arrayList->aList[i]);
    }

    strcpy(arrayList->aList[index], str);

    return 0;

}

// --------------------------------------------------------------------------

// Check if an item at a given index is equal to value
int isItemAtEqual(int index, char *str, strArrayList arrayList) {

    if (index > arrayList.length || index < 0) {
        printf("isItemAtEqual Error: Index is out of range\n");
        return -1;
    }

    if (strcmp(arrayList.aList[index], str) == 0) {
        return 1;
    }

    return 0;

}

// --------------------------------------------------------------------------

// Remove item at speficied index
int removeAtIndex(int index, strArrayList *arrayList) {

    int i;

    if (index > arrayList->length || index < 0) {
        printf("insertAtEnd Error: Index is out of range\n");
        return 1;
    }

    // Shift contents over
    for (i=index; i<arrayList->length; i++) {
        if (i+1 >= getMaxSize(*arrayList)) {
            break;
        }
        strcpy(arrayList->aList[i], arrayList->aList[i+1]);
    }

    arrayList->length--;

    return 0;

}

// Remove specified item from list (if it exists)
int removeItem(char *str, strArrayList *arrayList) {

    int i;
    int index;

    // Get index of value in the list
    index = searchArrayList(str, *arrayList);

    if (index == -1) {
        printf("removeItem Error: List does not contain '%s'.  Cannot remove item\n", str);
        return 1;
    }

    // Shift contents over
    for (i=index; i<arrayList->length; i++) {
        strcpy(arrayList->aList[i], arrayList->aList[i+1]);
    }

    arrayList->length--;

    return 0;

}

// --------------------------------------------------------------------------

// Remove specified item from list (if it exists)
int getItemAtIndex(int index, char *str, strArrayList arrayList) {

    if (index > arrayList.length || index < 0) {
        printf("getItemAtIndex Error: Index is out of range\n");
        return -1;
    }

    strcpy(str, arrayList.aList[index]);

    return 0;

}

// --------------------------------------------------------------------------

// Print the list contents
void printList(strArrayList arrayList) {

    int i;

    printf("String ArrayList:\n");

    for (i=0; i<arrayList.length; i++) {

        if (i>0 && i%10==10) {
            printf("\n");
        }

        printf("%s, ", arrayList.aList[i]);

    }

    printf("\n");
}

// --------------------------------------------------------------------------

// Free the dynamically allocated ArrayList
void freeList(strArrayList *arrayList) {
    int i;

    for (i=0; i<getMaxSize(*arrayList); i++) {
      free(arrayList->aList[i]);
    }

    free(arrayList->aList);
}

//-----------------------------------------------------------------------------
 int main() {


   char str[255];

   strArrayList strList1;

   StrArrayList(10,255, &strList1);

   insertAtEnd("This"  , &strList1);
   insertAtEnd("is"    , &strList1);
   insertAtEnd("my"    , &strList1);
   insertAtEnd("string", &strList1);
   insertAtEnd("array" , &strList1);
   insertAtEnd("list"  , &strList1);

   printList(strList1);

   printf("\nIs it empty? %d\n", isEmpty(strList1));

   printf("\n Let's insert 'IS' into the list at index 2\n");
   insertAtIndex(2, "IS", &strList1);
   printList(strList1);

   printf("\n Let's insert 'STR' into the list at index 4\n");
   insertAtIndex(4, "STR", &strList1);
   printList(strList1);

   printf("\n Let's insert 'STR' again into the list at index 4.  Should fail\n");
   insertAtIndex(4, "STR", &strList1);
   printList(strList1);

   printf("\n Let's remove the item at index 4\n");
   removeAtIndex(4, &strList1);
   printList(strList1);

   printf("\n Let's remove the item at index 40.  Should fail.\n");
   removeAtIndex(40, &strList1);
   printList(strList1);

   printf("\n Let's remove the item 'This' from the list.\n");
   removeItem("This", &strList1);
   printList(strList1);

   printf("\n Let's remove the item 'NotInList' from the list.  Should fail\n");
   removeItem("NotInList", &strList1);
   printList(strList1);

   printf("\nLet's get an item from the list\n");
   getItemAtIndex(3, str, strList1);
   printf("Item at index 3 is '%s'\n", str);

   freeList(&strList1);

 }

File: lists/arrayList/go/src/arrayList/StrArrayList.go
package arrayList

import "fmt"

type strListStruct struct {
  maxSize int
  length  int
  aList   []string
}

// ----------------------------------------------------------------------------

// Initializes the array list and its structure components and returns the
// initialized struct
func StrArrayList(aSize int) strListStruct {

  var list strListStruct
  var indx int

  // Initialize the array list structure elements
  list.maxSize = aSize
  list.length  = 0

  indx = 0
  for indx < aSize {
      list.aList   = append(list.aList, "")
      indx = indx+1
  }

  return list
}

// ----------------------------------------------------------------------------

// Returns if the list is is full
func StrArrayListIsFull(list strListStruct) bool {

  var result bool

  result = false

  if list.length == list.maxSize {
    result = true
  }
  return result

}

// ----------------------------------------------------------------------------

// Returns if the list is is empty
func StrArrayListIsEmpty(list strListStruct) bool {

  var result bool

  result = false

  if list.length == 0 {
    result = true
  }
  return result

}

// ----------------------------------------------------------------------------

// Returns the maximum size of the array list
func StrArrayListMaxSize(list strListStruct) int {
  return list.maxSize
}

// ----------------------------------------------------------------------------

// Returns the current length of the list
func StrArrayListLength(list strListStruct) int {
  return list.length
}

// ----------------------------------------------------------------------------

// Search for an item in the list and return the index if it exists
func SearchStrArrayList(value string, list strListStruct) int {

  var indx int

  for indx < list.length {
    if value == list.aList[indx] {
      return indx
    }
    indx++
  }

  return -1
}

// ----------------------------------------------------------------------------

// Adds an item at the end of the list.  The item cannot already exist and
// the list cannot be full.
func StrArrayListInsertAtEnd(value string, list *strListStruct) int {

  // Is the list full?
  if StrArrayListIsFull(*list) == true {
    fmt.Println("insertAtEnd Error: List is full cannot add new item")
    return 1
  }

  // Does the list already have the value
  if SearchStrArrayList(value, *list) > -1 {
    fmt.Println("insertAtEnd Error: List already has the value ", value)
    return 1
  }

  list.aList[list.length] = value
  list.length++

  return 0
}

// ----------------------------------------------------------------------------

// Adds an item into the list at the specified index.  If the list is full,
// or the value is already in the list or the index given is invalid then
// nothing is added
func StrArrayListInsertAtIndex(index int, value string, list *strListStruct) int {

  var i int

  // Is the list full?
  if StrArrayListIsFull(*list) == true {
    fmt.Println("insertAtEnd Error: List is full cannot add new item")
    return 1
  }

  // Is the index valid?
  if (index < 0 || index > list.length) {
    fmt.Println("insertAtEnd Error: Index given is outside the range of list")
    return 1
  }

  // Does the list already have the value
  if SearchStrArrayList(value, *list) > -1 {
    fmt.Println("insertAtEnd Error: List already has the value ", value)
    return 1
  }

  // Shift the contents over by one after index
  i = list.length
  for i > index {
    list.aList[i] = list.aList[i-1]
    i--
  }

  list.aList[index] = value
  list.length++

  return 0
}

// ----------------------------------------------------------------------------

// Removes the item at specified index.  The index must be valid
func StrArrayListRemoveAtIndex(index int, list *strListStruct) int {

  var i int

  // Is the index valid?
  if (index < 0 || index > list.length) {
    fmt.Println("insertAtEnd Error: Index given is outside the range of list")
    return 1
  }

  // Shift the contents over by one after index
  i = index
  for i < list.length-1 {
    list.aList[i] = list.aList[i+1]
    i++
  }

  list.length--

  return 0
}

// ----------------------------------------------------------------------------

// Searchs for a value in the list and removes it.  If the item does not exist
// then nothing is done.
func StrArrayListRemove(value string, list *strListStruct) int {

  var i     int
  var index int

  // Get the index of the value in the list.  If -1 is returned then the
  // value does not exist in the list
  index = SearchStrArrayList(value, *list)

  // Is the index valid?
  if (index == -1) {
    fmt.Println("Remove Error: Value ", value, " does not exist in list.  Cannot remove")
    return 1
  }

  // Shift the contents over by one after index
  i = index
  for i < list.length-1 {
    list.aList[i] = list.aList[i+1]
    i++
  }

  list.length--

  return 0
}

// ----------------------------------------------------------------------------

// Searchs for a value in the list and removes it.  If the item does not exist
// then nothing is done.  Returns a map which identifies if value was found
// and the value itself.  If not found then zero's are returned
func StrArrayListGetItemAtIndex(index int, list strListStruct) map[string]string {

  result := make(map[string]string)

  // Is the index valid?
  if (index < 0 || index > list.length) {
    fmt.Println("GetItemAtIndex Error: Index is not in valid range")
    result["valid"] = "0"
    result["value"] = ""
    return result
  }

  result["valid"] = "1"
  result["value"] = list.aList[index]

  return result
}

// ----------------------------------------------------------------------------

// Check if an item is equal at an index
func StrArrayListIsItemAtEqual(index int, value string, list strListStruct) bool {

  // Is the index valid?
  if (index < 0 || index > list.length) {
    fmt.Println("IsItemAtEqual Error: Index is not in valid range")
    return false
  }

  if list.aList[index] == value {
    return true
  }

  return false
}

// ----------------------------------------------------------------------------

// Print the contents of the array list
func PrintStrArrayList(list strListStruct) {

  var indx int

  fmt.Println("STRING ARRAY LIST: ")

  for indx < list.length {
    if indx % 10 == 0 && indx != 0 {
      fmt.Println("")
    }
    fmt.Print(list.aList[indx], ", ")
    indx = indx + 1
  }
  fmt.Println("")

}

// ----------------------------------------------------------------------------

// Clear the array list
func ClearStrArrayList(list *strListStruct) {
  list.length = 0
}
